openpgp-dsm/src/lib.rs

```diff
@@ -68,43 +77,122 @@ pub struct DsmAgent {
     role:        Role,
 }
 
-const ENV_API_KEY: &str = "FORTANIX_API_KEY";
+const DSM_LABEL_PGP:    &str = "sq_dsm";
+const ENV_API_KEY:      &str = "FORTANIX_API_KEY";
 const ENV_API_ENDPOINT: &str = "FORTANIX_API_ENDPOINT";
-const ENV_HTTP_PROXY: &str = "http_proxy";
-const ENV_NO_PROXY: &str = "no_proxy";
-const DSM_LABEL_PGP: &str = "sq_dsm";
-const MIN_DSM_VERSION: &str = "4.2.0";
+const ENV_APP_UUID:     &str = "FORTANIX_APP_UUID";
+const ENV_HTTP_PROXY:   &str = "http_proxy";
+const ENV_NO_PROXY:     &str = "no_proxy";
+const ENV_P12:          &str = "FORTANIX_PKCS12_ID";
+const MIN_DSM_VERSION:  &str = "4.2.0";
+
+#[derive(Clone)]
+enum Auth {
+    ApiKey(String),
+    // App UUID and PKCS12 identity
+    Cert(Uuid, Identity),
+}
 
-struct Credentials {
+#[derive(Clone)]
+pub struct Credentials {
     api_endpoint: String,
-    api_key:      String,
-    proxy:        Option<Arc<HyperClient>>,
+    auth:         Auth,
 }
 
 impl Credentials {
-    fn new_from_env() -> Result<Self> {
+    pub fn new_from_env() -> Result<Self> {
         let api_endpoint = env::var(ENV_API_ENDPOINT)
             .with_context(|| format!("{} env var absent", ENV_API_ENDPOINT))?;
-        let api_key = env::var(ENV_API_KEY)
-            .with_context(|| format!("{} env var absent", ENV_API_KEY))?;
-        let proxy = decide_proxy_from_env(&api_endpoint);
-
-        Ok(Self {
-            api_endpoint,
-            api_key,
-            proxy,
-        })
+
+        let auth = match (
+            env::var(ENV_API_KEY).ok(),
+            env::var(ENV_P12).ok(),
+        ) {
+            (Some(api_key), other) => {
+                if other.is_some() {
+                    println!("Both {}, {} are set, using API key auth",
```

Comment:
I see you are using the opposite convention I used for my CNG client changes :slightly_smiling_face:. But no matter...

(I don't think it's worth changing your code just so it matches the convention I used)